nbconvert/preprocessors/execute.py

```diff
@@ -197,8 +217,12 @@ def run_cell(self, cell):
         # wait for finish, with timeout
         while True:
             try:
-                timeout = self.timeout
-                if timeout < 0:
+                if self.timeout_func is not None:
+                    timeout = self.timeout_func(cell.source)
```

Comment:
Can you pass the cell itself, rather than cell.source? It's conceivable that one could write a timeout_func that looks at metadata, etc.